### PR TITLE
Add support for prepared statements

### PR DESCRIPTION
**Pull request by arybczak**

### CODE REVIEW

File: src/Database/PostgreSQL/PQTypes/Internal/Query.hs [hunk 1]
```diff
@@ -30,11 +35,26 @@ runQueryIO sql st = do
...
     -- are able to receive asynchronous exceptions (assuming that threaded GHC
     -- runtime system is used) and react appropriately.
-    queryRunner <- async . restore $ do
-      let allocParam = ParamAllocator $ withPGparam cdPtr
-      (paramCount, res) <- withSQL sql allocParam $ \param query -> (,)
-        <$> (fromIntegral <$> c_PQparamCount param)
-        <*> c_PQparamExec cdPtr nullPtr param query c_RESULT_BINARY
+    let allocParam = ParamAllocator $ withPGparam cdPtr
+    queryRunner <- async . restore $ withSQL sql allocParam $ \param query format -> do
+      let queryName = BS.concat
```
Comment (arybczak) [inline src/Database/PostgreSQL/PQTypes/Internal/Query.hs:40]: @marco44 does that satisfy you? I wanted to make it longer, but it looks like PostgreSQL only considers first 64 characters :thinking:

Comment (arybczak) [reply src/Database/PostgreSQL/PQTypes/Internal/Query.hs:40]: Never mind, the trick is to put the hash first.

Comment (marco44) [reply src/Database/PostgreSQL/PQTypes/Internal/Query.hs:40]: From what I see, you're trying to remove all the current mechanism to replace it with prepared queries all over the place, right ? I think it's overkill, and could even be dangerous, we could anyway still need the old mechanism as a fallback: prepared queries static plan is not something we'd want all the time. I was more thinking of simply being able to use this new mechanism for a few queries (like the top few that are executed a lot with a small runtime and we're sure have a static plan, like fetching a document). For seldom run queries, prepared queries can be worse than normal one (that's why we have the 2 mechanisms :) ) : you need more round trips on the network, and you'll be compiling 2 plans: the static one, and the one taking into account all the parameters, for the first few runs.

To answer the original question, the "name" type is 64 chars (that's a compile option, but the default is 64). So yes, 64 is good. But why not pass it a name as a parameter in the function call ? That would be much easier for me when looking at the logs (I'll probably see a bunch of EXECUTE this_string). In my plan, we were going to do this for a select few queries, those where execution is small compared to planning, for optimization, not remove all that is in place

File: test/Main.hs [hunk 2]
```diff
@@ -316,6 +316,26 @@ autocommitTest td = testCase "Autocommit mode works" .
     assertEqualEq "Other connection sees autocommited data" 1 n
   runQuery_ $ rawSQL "DELETE FROM test1_ WHERE a = $1" sint
 
+preparedStatementTest :: TestData -> Test
+preparedStatementTest td = testCase "Execution of prepared statements works" .
+                           runTestEnv td defaultTransactionSettings $ do
+  let i1 = 42 :: Int32
+  runPreparedQuery_ "select1" $ "SELECT" <?> i1
+  o1 <- fetchOne runIdentity
+  assertEqualEq "Results match" i1 o1
```
Comment (jsynacek) [inline test/Main.hs:325]: Shouldn't there be some sort of a check that verifies that the prepared statement actually exists in the database and is prepared? I'm not sure if that's even possible.

File: test/Main.hs [hunk 3]
```diff
@@ -316,6 +316,26 @@ autocommitTest td = testCase "Autocommit mode works" .
...
+  o2 <- fetchOne runIdentity
+  assertEqualEq "Results match" i2 o2
+
+  let i3 = "lalala" :: String
+  -- Changing parameter type in an already prepared statement shouldn't work.
+  o3 <- try $ runPreparedQuery_ "select1" $ "SELECT" <?> i3
+  case o3 of
+    Left DBException{} -> pure ()
+    Right r3           -> liftBase . assertFailure $ "Expected DBException, but got" <+> show r3
+
```
Comment (jsynacek) [inline test/Main.hs:338]: `runPreparedQuery01` is not covered by the tests.

**Review by jsynacek — CHANGES_REQUESTED** [review]
Looks good, minus the missing test case for `runPreparedQuery01`.

Overall, I think that the test suite should do a bit more, as in slightly more complicated queries. I just don't know which ones exactly would be good candidates. It feels a bit underwhelming for what the PR adds.

File: test/Main.hs (hunk 2 above)
Comment (marco44) [reply test/Main.hs:325]: https://www.postgresql.org/docs/11/view-pg-prepared-statements.html ?

Comment (arybczak) [reply test/Main.hs]: Added.

File: test/Main.hs (hunk 3 above)
Comment (arybczak) [reply test/Main.hs:347]: Right, but so isn't runQuery01 ;) I know, a somewhat poor excuse.

Comment (jsynacek) [reply test/Main.hs:347]: Yeah, very poor :grin:

**Review by jsynacek — APPROVED** [review]
Ok, I'm not going to be annoying with the tests anymore!

Looks good.

**Review by jsynacek — APPROVED** [review]

**Review by marco44 — APPROVED** [review]

**Pull request merged**